File: ranker.go
// The product. Turns an unordered pile of matches into an ordered one.
//
// Two rules are structural, not heuristic:
//  1. a file holding a definition-shaped match always sorts above files that do
//     not, whatever its path prior — a definition is never buried;
//  2. every score component is reproducible and dumpable via --explain.

package main

import (
	"math"
	"math/bits"
	"sort"
	"strings"
	"unicode"
)

// Words that introduce a definition. Kept deliberately small: the backward scan
// below is what gives precision, not the size of this list.
var defKW = map[string]bool{
	"function": true, "fn": true, "def": true, "class": true, "interface": true,
	"type": true, "enum": true, "struct": true, "impl": true, "trait": true,
	"const": true, "let": true, "var": true, "export": true, "default": true,
	"public": true, "private": true, "protected": true, "static": true,
	"async": true, "func": true, "module": true, "namespace": true,
	"abstract": true, "override": true, "val": true, "object": true,
	"record": true, "protocol": true, "extension": true, "component": true,
	"procedure": true, "method": true, "macro_rules": true, "declare": true,
	"readonly": true,
}

// Characters that break the link between a definition keyword and an
// identifier. `=` is what stops `const [rows, setRows] = useState(...)` from
// being read as a definition of `useState`; `[` `]` `,` are deliberately absent
// so destructuring declarations (`const [rows, setRows] = …` seen from
// `setRows`) are still recognised.
const breakers = "=(){};:?|&<>"

func isIdentChar(c rune) bool {
	return unicode.IsLetter(c) || unicode.IsDigit(c) || c == '_' || c == '$'
}

type span struct{ start, end int }

// Byte offsets of every whole-word occurrence of any token in `line`.
func tokenPositions(line string, tokens []string) []span {
	lower := strings.ToLower(line)
	b := []byte(lower)
	var out []span
	for _, t := range tokens {
		if len(t) < 2 {
			continue
		}
		from := 0
		for from < len(lower) {
			rel := strings.Index(lower[from:], t)
			if rel < 0 {
				break
			}
			start := from + rel
			end := start + len(t)
			beforeOK := start == 0 || !isIdentChar(rune(b[start-1]))
			afterOK := end >= len(b) || !isIdentChar(rune(b[end]))
			if beforeOK && afterOK {
				out = append(out, span{start, end})
			}
			from = start + 1
		}
	}
	return out
}

// Walks backwards from an identifier looking for a definition keyword, stopping
// at the first breaker character.
func backwardDef(line string, start int) bool {
	if start > len(line) {
		start = len(line)
	}
	head := []rune(line[:start])
	var word []rune
	for i := len(head) - 1; i >= 0; i-- {
		c := head[i]
		if isIdentChar(c) {
			word = append([]rune{c}, word...)
			continue
		}
		if len(word) > 0 {
			if defKW[strings.ToLower(string(word))] {
				return true
			}
			word = nil
		}
		if strings.ContainsRune(breakers, c) {
			return false
		}
	}
	return len(word) > 0 && defKW[strings.ToLower(string(word))]
}

func firstWordIsDefKW(line string) bool {
	t := strings.TrimLeftFunc(line, unicode.IsSpace)
	var w strings.Builder
	for _, c := range t {
		if !isIdentChar(c) {
			break
		}
		w.WriteRune(c)
	}
	return defKW[strings.ToLower(w.String())]
}

func isComment(line string) bool {
	t := strings.TrimLeftFunc(line, unicode.IsSpace)
	return strings.HasPrefix(t, "//") || strings.HasPrefix(t, "#") ||
		strings.HasPrefix(t, "* ") || strings.HasPrefix(t, "/*") ||
		strings.HasPrefix(t, "--") || strings.HasPrefix(t, "<!--")
}

func isImport(line string) bool {
	t := strings.TrimLeftFunc(line, unicode.IsSpace)
	return strings.HasPrefix(t, "import ") || strings.HasPrefix(t, "from ") ||
		strings.HasPrefix(t, "#include") || strings.HasPrefix(t, "use ") ||
		strings.HasPrefix(t, "require(") || strings.HasPrefix(t, "export * ") ||
		strings.HasPrefix(t, "export {")
}

// Scores one match line and says whether it looks like a definition.
func lineScore(line string, tokens []string) (float32, bool) {
	score := float32(1.0)
	positions := tokenPositions(line, tokens)
	wholeWord := len(positions) > 0

	isDef := false
	for _, p := range positions {
		if backwardDef(line, p.start) {
			isDef = true
			break
		}
		after := ""
		if p.end <= len(line) {
			after = strings.TrimLeftFunc(line[p.end:], unicode.IsSpace)
		}
		followedByParen := strings.HasPrefix(after, "(") || strings.HasPrefix(after, "<")
		// An assignment before the identifier means it is being *called*, not
		// declared: `const [rows, setRows] = useState(...)` defines rows, not
		// useState. Without this guard every React hook call reads as a
		// definition — measured: 296 false positives on a single query.
		assignedTo := strings.Contains(line[:min(p.start, len(line))], "=")
		if followedByParen && !assignedTo {
			if firstWordIsDefKW(line) {
				isDef = true
				break
			}
			// C-style / Go / class methods: `int main() {`, `handleClick() {`
			var before rune = 0
			if p.start > 0 {
				r := []rune(line[:p.start])
				if len(r) > 0 {
					before = r[len(r)-1]
				}
			}
			callLike := before == '.' || before == '>'
			if !callLike && strings.HasSuffix(strings.TrimRightFunc(line, unicode.IsSpace), "{") {
				isDef = true
				break
			}
		}
	}

	if isDef {
		score += 3.0
	}
	if wholeWord {
		score += 0.6
	}
	if runeLen(line) < 120 {
		score += 0.3
	}
	if isImport(line) {
		score -= 0.9
	}
	if isComment(line) {
		score -= 0.6
	}
	return score, isDef
}

type Explain struct {
	Line, Coverage, Density, PathPrior, NameBonus, Dedup float32
}

func pathPrior(path string) float32 {
	// Leading slash so the segment patterns below also match at path start.
	p := "/" + strings.ToLower(path)
	junk := []string{
		"/node_modules/", "/dist/", "/build/", "/.next/", "/out/", "/vendor/",
		"/coverage/", "/target/", ".min.", "/generated/", "/.venv/", "/__pycache__/",
	}
	testy := []string{
		"/test/", "/tests/", "/spec/", "__tests__", "__mocks__", ".test.", ".spec.",
		"/e2e/", "/cypress/", "/fixtures/", "/mocks/", ".stories.", "/storybook/",
	}
	for _, j := range junk {
		if strings.Contains(p, j) {
			return 0.10
		}
	}
	for _, t := range testy {
		if strings.Contains(p, t) {
			return 0.30
		}
	}
	if strings.HasSuffix(p, ".d.ts") {
		return 0.50
	}
	if strings.Contains(p, "/src/") || strings.Contains(p, "/app/") || strings.Contains(p, "/lib/") {
		return 1.10
	}
	return 1.0
}

func nameBonus(path string, tokens []string) float32 {
	file := strings.ToLower(path)
	if i := strings.LastIndex(file, "/"); i >= 0 {
		file = file[i+1:]
	}
	stem := file
	if i := strings.Index(stem, "."); i >= 0 {
		stem = stem[:i]
	}
	stem = strings.NewReplacer("-", "", "_", "").Replace(stem)
	dir := strings.ToLower(path)
	for _, t := range tokens {
		if len(t) < 3 {
			continue
		}
		if strings.Contains(stem, t) {
			return 1.5
		}
	}
	for _, t := range tokens {
		if len(t) >= 4 && strings.Contains(dir, t) {
			return 0.7
		}
	}
	return 0.0
}

// Ranks files in place and returns the per-file score breakdown.
func rank(files []*FileHit, tokens []string, nTokens int) map[string]Explain {
	// Boilerplate detection: a representative line repeated across many files
	// (a shared import, a generated header) carries little information.
	repeats := map[string]int{}
	for _, f := range files {
		if len(f.Best) > 0 {
			repeats[f.Best[0].Text]++
		}
	}

	explains := map[string]Explain{}
	for _, f := range files {
		var line float32
		if f.Def != nil {
			line = f.Def.Score
		} else if len(f.Best) > 0 {
			line = f.Best[0].Score
		}
		// Coverage dominates on multi-term queries: a file mentioning four of
		// the five words asked about is what the agent wanted, even if each word
		// appears once. On a single-token query this term is constant.
		multi := nTokens > 1
		var coverage, density float32
		if multi {
			coverage = 3.0 * (float32(bits.OnesCount32(f.Covered)) / float32(nTokens))
		} else {
			// On a multi-term query raw match count is not relevance: a page
			// repeating a word 469 times outranked the actual module. What
			// discriminates there is the filename, not the volume.
			// f32 throughout, to match the reference implementation's rounding.
			density = float32(0.35 * float32(math.Log(float64(float32(1.0+float32(f.Matches))))))
			if density > 1.2 {
				density = 1.2
			}
		}
		prior := pathPrior(f.Path)
		// Filename match is the strongest signal available when the query is
		// prose, so it carries more weight there.
		name := nameBonus(f.Path, tokens)
		if multi {
			name *= 2.0
		}
		dedup := float32(1.0)
		if len(f.Best) > 0 && !f.Best[0].IsDef && repeats[f.Best[0].Text] > 3 {
			dedup = 0.6
		}
		f.Score = (line + coverage + density + name) * prior * dedup
		explains[f.Path] = Explain{line, coverage, density, prior, name, dedup}
	}

	// Hard key first: a file holding a definition never sorts below one that
	// does not. Path priors only order files within each group.
	sort.SliceStable(files, func(i, j int) bool {
		a, b := files[i], files[j]
		ad, bd := a.Def != nil, b.Def != nil
		if ad != bd {
			return ad
		}
		if a.Score != b.Score {
			return a.Score > b.Score
		}
		return a.Path < b.Path
	})
	return explains
}

File: src/ranker.rs
//! The product. Turns an unordered pile of matches into an ordered one.
//!
//! Two rules are structural, not heuristic:
//!   1. a file holding a definition-shaped match always sorts above files that
//!      do not, whatever its path prior — a definition is never buried;
//!   2. every score component is reproducible and dumpable via `--explain`.

use std::collections::HashMap;

use crate::executor::FileHit;

/// Words that introduce a definition. Kept deliberately small: the backward
/// scan below is what gives precision, not the size of this list.
const DEF_KW: &[&str] = &[
    "function", "fn", "def", "class", "interface", "type", "enum", "struct",
    "impl", "trait", "const", "let", "var", "export", "default", "public",
    "private", "protected", "static", "async", "func", "module", "namespace",
    "abstract", "override", "val", "object", "record", "protocol", "extension",
    "component", "procedure", "method", "macro_rules", "declare", "readonly",
];

/// Characters that break the link between a definition keyword and an
/// identifier. `=` is what stops `const [rows, setRows] = useState(...)` from
/// being read as a definition of `useState`; `[` `]` `,` are deliberately
/// absent so destructuring declarations (`const [rows, setRows] = …` seen from
/// `setRows`) are still recognised.
const BREAKERS: &[char] = &['=', '(', ')', '{', '}', ';', ':', '?', '|', '&', '<', '>'];

fn is_ident_char(c: char) -> bool {
    c.is_alphanumeric() || c == '_' || c == '$'
}

/// Byte offsets of every whole-word occurrence of any token in `line`.
fn token_positions(line: &str, tokens: &[String]) -> Vec<(usize, usize)> {
    let lower = line.to_lowercase();
    let bytes = lower.as_bytes();
    let mut out = Vec::new();
    for t in tokens {
        if t.len() < 2 {
            continue;
        }
        let mut from = 0usize;
        while let Some(rel) = lower[from..].find(t.as_str()) {
            let start = from + rel;
            let end = start + t.len();
            let before_ok = start == 0 || !is_ident_char(bytes[start - 1] as char);
            let after_ok = end >= bytes.len() || !is_ident_char(bytes[end] as char);
            if before_ok && after_ok {
                out.push((start, end));
            }
            from = start + 1;
            if from >= lower.len() {
                break;
            }
        }
    }
    out
}

/// Walks backwards from an identifier looking for a definition keyword,
/// stopping at the first breaker character.
fn backward_def(line: &str, start: usize) -> bool {
    let head = &line[..start];
    let mut word = String::new();
    for c in head.chars().rev() {
        if is_ident_char(c) {
            word.insert(0, c);
            continue;
        }
        if !word.is_empty() {
            if DEF_KW.contains(&word.to_lowercase().as_str()) {
                return true;
            }
            word.clear();
        }
        if BREAKERS.contains(&c) {
            return false;
        }
    }
    !word.is_empty() && DEF_KW.contains(&word.to_lowercase().as_str())
}

fn first_word_is_def_kw(line: &str) -> bool {
    let w: String = line.trim_start().chars().take_while(|c| is_ident_char(*c)).collect();
    DEF_KW.contains(&w.to_lowercase().as_str())
}

fn is_comment(line: &str) -> bool {
    let t = line.trim_start();
    t.starts_with("//") || t.starts_with('#') || t.starts_with("* ") || t.starts_with("/*")
        || t.starts_with("--") || t.starts_with("<!--")
}

fn is_import(line: &str) -> bool {
    let t = line.trim_start();
    t.starts_with("import ") || t.starts_with("from ") || t.starts_with("#include")
        || t.starts_with("use ") || t.starts_with("require(") || t.starts_with("export * ")
        || t.starts_with("export {")
}

/// Scores one match line and says whether it looks like a definition.
pub fn line_score(line: &str, tokens: &[String]) -> (f32, bool) {
    let mut score = 1.0f32;
    let positions = token_positions(line, tokens);
    let whole_word = !positions.is_empty();

    let mut is_def = false;
    for (start, end) in &positions {
        if backward_def(line, *start) {
            is_def = true;
            break;
        }
        let after = line[*end..].trim_start();
        let followed_by_paren = after.starts_with('(') || after.starts_with('<');
        // An assignment before the identifier means it is being *called*, not
        // declared: `const [rows, setRows] = useState(...)` defines rows, not
        // useState. Without this guard every React hook call reads as a
        // definition — measured: 296 false positives on a single query.
        let assigned_to = line[..*start].contains('=');
        if followed_by_paren && !assigned_to {
            if first_word_is_def_kw(line) {
                is_def = true;
                break;
            }
            // C-style / Go / class methods: `int main() {`, `handleClick() {`
            let before = line[..*start].chars().last();
            let call_like = matches!(before, Some('.') | Some('>'));
            if !call_like && line.trim_end().ends_with('{') {
                is_def = true;
                break;
            }
        }
    }

    if is_def {
        score += 3.0;
    }
    if whole_word {
        score += 0.6;
    }
    if line.chars().count() < 120 {
        score += 0.3;
    }
    if is_import(line) {
        score -= 0.9;
    }
    if is_comment(line) {
        score -= 0.6;
    }
    (score, is_def)
}

#[derive(Debug, Default, Clone)]
pub struct Explain {
    pub line: f32,
    pub coverage: f32,
    pub density: f32,
    pub path_prior: f32,
    pub name_bonus: f32,
    pub dedup: f32,
}

fn path_prior(path: &str) -> f32 {
    // Leading slash so the segment patterns below also match at path start.
    let p = format!("/{}", path.to_lowercase());
    let p = p.as_str();
    const JUNK: &[&str] = &[
        "/node_modules/", "/dist/", "/build/", "/.next/", "/out/", "/vendor/",
        "/coverage/", "/target/", ".min.", "/generated/", "/.venv/", "/__pycache__/",
    ];
    const TESTY: &[&str] = &[
        "/test/", "/tests/", "/spec/", "__tests__", "__mocks__", ".test.", ".spec.",
        "/e2e/", "/cypress/", "/fixtures/", "/mocks/", ".stories.", "/storybook/",
    ];
    if JUNK.iter().any(|j| p.contains(j)) {
        return 0.10;
    }
    if TESTY.iter().any(|t| p.contains(t)) {
        return 0.30;
    }
    if p.ends_with(".d.ts") {
        return 0.50;
    }
    if p.contains("/src/") || p.contains("/app/") || p.contains("/lib/") {
        return 1.10;
    }
    1.0
}

fn name_bonus(path: &str, tokens: &[String]) -> f32 {
    let file = path.rsplit('/').next().unwrap_or(path).to_lowercase();
    let stem = file.split('.').next().unwrap_or(&file).replace(['-', '_'], "");
    let dir = path.to_lowercase();
    for t in tokens {
        if t.len() < 3 {
            continue;
        }
        if stem.contains(t.as_str()) {
            return 1.5;
        }
    }
    for t in tokens {
        if t.len() >= 4 && dir.contains(t.as_str()) {
            return 0.7;
        }
    }
    0.0
}

/// Ranks files in place and returns the per-file score breakdown.
pub fn rank(files: &mut [FileHit], tokens: &[String], n_tokens: usize) -> HashMap<String, Explain> {
    // Boilerplate detection: a representative line repeated across many files
    // (a shared import, a generated header) carries little information.
    let mut seen: HashMap<&str, u32> = HashMap::new();
    for f in files.iter() {
        if let Some(l) = f.best.first() {
            *seen.entry(l.text.as_str()).or_insert(0) += 1;
        }
    }
    let repeats: HashMap<String, u32> =
        seen.into_iter().map(|(k, v)| (k.to_string(), v)).collect();

    let mut explains = HashMap::new();
    for f in files.iter_mut() {
        let line = f
            .def
            .as_ref()
            .map(|l| l.score)
            .unwrap_or_else(|| f.best.first().map(|l| l.score).unwrap_or(0.0));
        // Coverage dominates on multi-term queries: a file mentioning four of
        // the five words asked about is what the agent wanted, even if each
        // word appears once. On a single-token query this term is constant and
        // changes nothing.
        let multi = n_tokens > 1;
        let coverage = if multi {
            3.0 * (f.covered.count_ones() as f32 / n_tokens as f32)
        } else {
            0.0
        };
        // On a multi-term query raw match count is not relevance: an SEO page
        // saying "vendeur" 469 times outranked the actual quota module. What
        // discriminates there is the filename, not the volume.
        let density = if multi {
            0.0
        } else {
            (0.35 * (1.0 + f.matches as f32).ln()).min(1.2)
        };
        let prior = path_prior(&f.path);
        // Filename match is the strongest signal available when the query is
        // prose, so it carries more weight there.
        let name = name_bonus(&f.path, tokens) * if multi { 2.0 } else { 1.0 };
        let dedup = match f.best.first() {
            Some(l) if !l.is_def && repeats.get(&l.text).copied().unwrap_or(0) > 3 => 0.6,
            _ => 1.0,
        };
        f.score = (line + coverage + density + name) * prior * dedup;
        explains.insert(
            f.path.clone(),
            Explain { line, coverage, density, path_prior: prior, name_bonus: name, dedup },
        );
    }

    // Hard key first: a file holding a definition never sorts below one that
    // does not. Path priors only order files within each group.
    files.sort_by(|a, b| {
        b.def
            .is_some()
            .cmp(&a.def.is_some())
            .then_with(|| b.score.total_cmp(&a.score))
            .then_with(|| a.path.cmp(&b.path))
    });
    explains
}

#[cfg(test)]
mod tests {
    use super::*;

    fn toks(q: &str) -> Vec<String> {
        crate::planner::tokenize(q)
    }

    #[test]
    fn call_is_not_a_definition() {
        let t = toks("useState");
        // Regression: these three shapes produced 296 false `[def]` marks on a
        // single query before the assignment guard.
        for line in [
            "const [rows, setRows] = useState<StockRow[]>([]);",
            "  const [inputMode, setInputMode] = React.useState<\"plate\" | \"manual\">(\"plate\")",
            "    const [isMobile, setIsMobile] = React.useState<boolean | undefined>(undefined)",
            "  setCount(useState(0));",
        ] {
            let (_, d) = line_score(line, &t);
            assert!(!d, "a useState() call must not be read as a definition: {line}");
        }
    }

    #[test]
    fn definition_of_the_queried_symbol_still_detected() {
        // The same shape *is* a definition when the query names the thing
        // being declared rather than the thing being called.
        let (_, d) = line_score(
            "  const [rows, setRows] = useState<StockRow[]>([]);",
            &toks("setRows"),
        );
        assert!(d, "setRows is genuinely declared on this line");
    }

    #[test]
    fn definitions_across_languages() {
        let cases = [
            ("handleSubmit", "export const handleSubmit = async (e: Event) => {"),
            ("handleSubmit", "export default function handleSubmit(e) {"),
            ("UserService", "export class UserService implements IUserService {"),
            ("Handle", "func (s *Server) Handle(w http.ResponseWriter, r *http.Request) {"),
            ("main", "int main(int argc, char **argv) {"),
            ("parse_config", "pub(crate) fn parse_config(path: &Path) -> Result<Config> {"),
            ("get_user", "    def get_user(self, user_id: int) -> User:"),
            ("Repository", "type Repository interface {"),
            ("handleClick", "  handleClick() {"),
            ("Config", "pub struct Config {"),
        ];
        for (q, line) in cases {
            let (_, d) = line_score(line, &toks(q));
            assert!(d, "should detect a definition in: {line}");
        }
    }

    #[test]
    fn imports_rank_below_uses() {
        let t = toks("useState");
        let (imp, _) = line_score("import { useState } from 'react';", &t);
        let (usage, _) = line_score("  const [a, setA] = useState(null);", &t);
        assert!(usage > imp, "an import line must not outrank a real use");
    }

    #[test]
    fn french_prose_is_not_a_definition_source() {
        // `type de vendeur` put SEO copy at the top of a natural-language
        // query because `type` is a definition keyword. Multi-term queries now
        // disable definition detection entirely; this test pins the shape that
        // exposed it.
        let (_, d) = line_score("<p>La protection légale dépend du type de vendeur :</p>", &toks("vendeur"));
        assert!(d, "the backward scan does fire here — which is why plan_multi turns it off");
    }

    #[test]
    fn junk_paths_are_demoted() {
        assert!(path_prior("node_modules/react/index.js") < path_prior("apps/web/src/a.ts"));
        assert!(path_prior("src/a.test.ts") < path_prior("src/a.ts"));
    }
}
